### DISCUSSION TITLE
compilation problems

### DISCUSSION THREAD

**Discussion by michaelsarr** [picocomputer/community#24] [Q&A] [answered]
i got the circuit board in the mail yesterday, soldered it up and it comes up and seems ok.

so on ubuntu  i have vscode installed  with the vscode template for hello world. It builds without errors.

but no luck with  ctrl - shift - B    im getting a permission denied error when trying to open dev/ttyACM0
it have connected to the serial connection with minicom
the help and status seem to work via the serial so not sure but i can play around with that.

I also tried taking the hello binary and putting it on a thumb drive  
i can see the drive in status , change to that drive . use ls and see the files there.

and when I try to use "load hello" i get Missing RP6502 ROM header

does this mean that it is running on the 6502 or is this interface running on the pico

so I just put this thing together. not sure if it actually works... i can read and write to memory
are there some downloadable binaries to test with somewhere? 

At this point im not sure if there is something wrong with the device or my compilation setup is not correct.

**Comment by rumbledethumps** [comment]
Permission denied is coming from Linux. Focus on that.

**Comment by brentward** [comment, marked as answer]
The permissions error you are getting on the /dev/ttyACM0 device could be because your user isn't in the dialout group. You could try adding your user to that group. Do `ls /dev/ttyACM0` and see what the ownership is and add your user to that group.

Once you get permission issue fixed you will be able to use the underlying rp6502.py file to upload the binary to a USB drive mounted to the RP6502. When you upload it through the script the headers get added to the binary file that allow you to install it as a ROM. I only had success when I manually specified the starting address. This is the command I used to upload the hello world file: 
`rp6502-sdk/rp6502.py -a 0x200 upload build/hello `

Once it is uploaded you will see the hello file when you do an LS on the RP6502 itself and you can install it using the command:
`INSTALL HELLO`
After that you will see it listed as one of the installed ROMs.

For more information examine how the basic.py file works inside the picocomputer/ehbasic repository.

**Reply by michaelsarr** [reply]
i just did a chmod on /dev/ttyACM0 and was able to push the hello world from VScode and run it on the pico computer..
so im glad to see it run some code. just wasnt sure if there was a bigger issue

I did try to copy the video of doing hello world in assembly but when I tried to start it with jmp $200 and i got unknown command I was getting a bit worried that there may be a larger issue.. thanks for the feedback

**Reply by mklcolvin** [reply]
I'm running into the same 'Missing RP6502 ROM Header' error, but I'm not connected to the RP itself, instead I'm just compiling on my Mac using VSC, copying the hello binary to a USB drive, and connecting that USB drive to the RP.  From what I'm reading from you, the rp6502.py script wraps the hello binary with the proper headings before sending it out to the RP over a physical connection?  It looks like the only options to the script are 'run' or 'upload' - is there a way to get it to re-write out the binary with the header info onto the source drive?  I'm not a python person (yet), so I'm still groping around this.

**Reply by rumbledethumps** [reply]
The rp6502.py tool doesn't yet have any local actions. It should do what you need in the future, but right now it does everything over the serial connection.

**Comment by mklcolvin** [comment]
I worked some on modifying the rp6502.py tool, and it now outputs a file that I think has the ROM header info, but I don't know what is a proper header. I'm attaching the hello world output from the template to show what it now outputs to a local file. If someone can post an actual binary and start address, I might be able to figure this out.

[hello.rom.zip](https://github.com/picocomputer/discussions/files/12772768/hello.rom.zip)

**Reply by netzerohero** [reply]
@mklcolvin - I too have been "groping around this" host-to-rp6502 object-file/ROM-file workflow; and have been reverse engineering the examples, vscode scaffolding and python scripts.

I looked at your hello.rom file contents; it doesn't contain any binary object code; your file only contains ascii of the assembly source with the appended header info.  So there's an issue to resolve.

Attached is a working hello world object binary with a "ROM'ed" version. Use your favorite binary-editor to view their contents and structure.  Zipfile also has some readme notes. 

[hello_rp6502.zip](https://github.com/picocomputer/discussions/files/13197065/hello_rp6502.zip)
  
To aid me, I made a slight modification to my local rp6502.py script to insert more helpful comment-info for the INFO monitor command:

```
    # TODO make this a generic upload
    if (args.command == 'upload'):
        rom = ROM()
 --->  rom.comment("# ROMFile: " + args.out + " Start addrs: " + hex(args.address))
        rom.binary_file(args.filename, args.address)
        rom.reset_vector(args.reset)
        mon.upload(args.out or os.path.basename(args.filename), rom)
```
]LS
     137 hello.rp6502                                                                                                                   

]INFO HELLO.RP6502                                                                                                                    
 "# ROMFile: hello.rp6502 Start addrs: 0x200" 


Hope this helps.

**Reply by rumbledethumps** [reply]
An entirely new asset workflow is being worked on right now. Wait for announcement before putting a lot of effort in the current one, if you can. Work in progress here: https://github.com/picocomputer/rp6502-sdk/tree/assets

**Reply by netzerohero** [reply]
Appreciate the insight & comment; current effort for me is pleasure in mixing the old-ways along with learning the new.  Thank you for creating a great project with elements of both!  Will be looking forward to the announcement. :)

**Reply by rumbledethumps** [reply]
Asset management in the SDK
https://www.patreon.com/posts/91883985

**Reply by mklcolvin** [reply]
Just got back in town, so I can take another look at this - many thanks to both of you!  Can't wait to try out the new asset management workflow!
